### ISSUE TITLE
Remove hard ArgoCD dependency

### ISSUE THREAD

**Issue by AndersBennedsgaard** [deployKF/deployKF#110]
### Checks

- [X] I have searched the [existing issues](https://github.com/deployKF/deployKF).
- [X] This request is NOT specific to the CLI. (If so, please open an issue on the [CLI repo](https://github.com/deployKF/cli/issues))

### Motivation

ArgoCD is a fine tool deploy stuff to a Kubernetes cluster, but not everybody want to use it (they could be using a different GitOps engine such as Flux, Fleet, etc., or not wanting to use GitOps).
I think it is fine to keep it as a recommendation, but we should not do what KubeFlow has done, and add arbitrary dependencies to their platform which aren't actually needed for a lot of people (in my mind, their implementation of Istio is this is the biggest contributor to KubeFlow being hard to install and manage)

### Are you willing & able to help?

- [ ] I am able to submit a PR!
- [ ] I can help test the feature!

**Comment by thesuperzapper** [comment]
@AndersBennedsgaard While I agree that it would be nice to use deployKF without ArgoCD, there are a few things that [deployKF uses ArgoCD for](https://www.deploykf.org/guides/dependencies/argocd/#how-does-deploykf-use-argo-cd), which can't be solved when using a raw manifests directory:

1. __Cleaning up resources:__
     - If you just manually apply Kustomize manifests to a Kubernetes cluster, there is no easy way to remove them again once they are no longer needed.
     - Obviously, this is less important for Helm apps, but deployKF includes both Helm and Kustomize apps (notably Kubeflow, lol).
2. __Syncing in a controlled way:__
      - We have a reasonably complex [sync process](https://github.com/deployKF/deployKF/blob/main/scripts/sync_argocd_apps.sh) which uses ArgoCD to ensure that each app is properly synced before moving on to the next.
      - Within many apps, we make use of the [`argocd.argoproj.io/sync-wave`](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) feature, to prevent race conditions _(e.g. to ensure [`cert-manager` is synced and healthy before syncing `ClusterIssuer` resources](https://github.com/deployKF/deployKF/blob/v0.1.4/generator/templates/manifests/deploykf-dependencies/cert-manager/templates/ClusterIssuer-kubeflow-gateway-issuer.yaml#L9))_
      - Many of our apps use [`helm.sh/hook: pre-install,pre-upgrade`](https://helm.sh/docs/topics/charts_hooks/) hooks _(e.g. to [generate unique OIDC secrets](https://github.com/deployKF/deployKF/blob/v0.1.4/generator/templates/manifests/deploykf-core/deploykf-auth/templates/generate/Job-generate-secrets.yaml))_ these take a non-trivial amount of time. Using ArgoCD allows us to only run them when their app is "out of sync" which we could not possibly know without ArgoCD.

Given the complexity of achieving all this natively in the `deploykf` CLI, and the significantly lower popularity of FluxCD, it's just really not a priority at the moment (and I am sure you would prefer I spend my time adding user-facing features!)
__Do you have a specific thing which is driving your desire to not use ArgoCD?__

---

_PS: from my discussions with large companies all around the world, pretty much everyone who is using Kubernetes seriously is using ArgoCD. For this and many other reasons, ArgoCD is how I recommend anyone deploy something as complex as deployKF. Yes, it has its problems, but for better or worse, it is the standard._

**Comment by AndersBennedsgaard** [comment]
Actually, no, I wouldn't prefer that you spend time on user-facing features. _I_ would prefer that the focus is on the stability and availability of use. But you probably prefer to focus on user-facing features as it is just more fun to work on :wink: 

But I have to disagree with several things you write.

> If you just manually apply Kustomize manifests to a Kubernetes cluster, there is no easy way to remove them again once they are no longer needed.

While I agree that a good way of removing stuff should be possible, is is very easy to just write `kubectl delete -f /path/to/stuff` with a potential `-R` for recursively deleting stuff. I am not advocating for using raw `kubectl` commands (even automated commands using Jenkins or similar); I am just saying that without the raw manifests, you remove a lot of potential users from using your product (which is fine if that is the decision).

> We have a reasonably complex [sync process](https://github.com/deployKF/deployKF/blob/main/scripts/sync_argocd_apps.sh) which uses ArgoCD to ensure that each app is properly synced before moving on to the next.

As I read it, this script is making sure that all the manifests are correctly applied to the cluster, in the correct order, and verifying that some resources has been pruned?
If that is the case: that is basically a solved problem in Flux. Here you just commit stuff to git, and specify which "folders"/"paths" are depending on each other. Then Flux will try to apply the contents of one path before moving on to the next.
I would be surprised if ArgoCD and other similar tools doesn't have this feature built in?

> Within many apps, we make use of the [argocd.argoproj.io/sync-wave](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) feature, to prevent race conditions (e.g. to ensure [cert-manager is synced and healthy before syncing ClusterIssuer resources](https://github.com/deployKF/deployKF/blob/v0.1.4/generator/templates/manifests/deploykf-dependencies/cert-manager/templates/ClusterIssuer-kubeflow-gateway-issuer.yaml#L9))

Same as above

> Many of our apps use [helm.sh/hook: pre-install,pre-upgrade](https://helm.sh/docs/topics/charts_hooks/) hooks (e.g. to [generate unique OIDC secrets](https://github.com/deployKF/deployKF/blob/v0.1.4/generator/templates/manifests/deploykf-core/deploykf-auth/templates/generate/Job-generate-secrets.yaml)) these take a non-trivial amount of time. Using ArgoCD allows us to only run them when their app is "out of sync" which we could not possibly know without ArgoCD.

This is just a Helm feature, so ArgoCD isn't really needed for this?

> Do you have a specific thing which is driving your desire to not use ArgoCD?

Yes. I've used ArgoCD, and I am not very impressed. It has a very nice developer-friendly UI, and if you don't do pure GitOps then it's a very nice alternative to `kubectl`. But as I am an infrastructure guy for the most part, I require stability when working with mutating webhooks, CRDs, dependencies to other stuff, and so on, which ArgoCD does not handle well. As I also want to use GitOps, I do not want to have state in both ArgoCDs database, git, and Kubernetes at the same time. I only want it to be in git, which ArgoCD _can_ do with read-only mode, but it is janky at best.
All of this is solved by Flux, and other GitOps engines. 

> PS: from my discussions with large companies all around the world, pretty much everyone who is using Kubernetes seriously is using ArgoCD. For this and many other reasons, ArgoCD is how I recommend anyone deploy something as complex as deployKF. Yes, it has its problems, but for better or worse, it is the standard.

While it is probably true that ArgoCD is the most used engine (I am not entirely sure that it is, since a lot of people use them in conjunction), I've worked with a lot of different companies in my role as consultant, and I've only seen the use of ArgoCD once - when I wanted to test it on one of our demo setups.
It might be that I am working in a more security-heavy field where we require an audit-trail for all changes (mostly in the form of the git history), or something else, but as I see it (and mentioned above), choosing ArgoCD as the only way to support your tool will not make it usable for a LOT of people.
Which might be fine. In that case, feel free to close this issue.

**Comment by thesuperzapper** [comment]
@AndersBennedsgaard my comment in https://github.com/deployKF/deployKF/issues/110#issuecomment-1968098537 was primarily about why raw manifests wouldn't work. However, it's always been a roadmap item to support FluxCD (given it's the only other widely used K8S CD tool).

If you or your customers would benefit from FluxCD support coming sooner, I am very happy to help facilitate such a contribution from you or your team (or others watching this thread).

**Comment by AndersBennedsgaard** [comment]
I am not actually arguing for full-blown Flux support, but rather removal of the hard ArgoCD dependency. It's fair to have support for ArgoCD, Flux, whatever other system using plugins or some other ways, but there are times where raw manifests are just preferred (such as when using Flux).
Unfortunately, I don't have the most time in the world right now, but if I do find some I would love to take a stab at it

**Comment by Zeratoxx** [comment]
Is a Kubernetes Operator an option?

**Comment by mnasty** [comment]
I would like to second this request. As a lead engineer on an ML team functioning like a startup and provisioning our own siloed infrastructure for compliance reasons, it really sucks to have to do a complete deployment of a separate product just for a single deployment of something unrelated..

I understand that this is a *very* complex deployment but focusing on Kubernetes native solutions to the challenges makes your product more accessible. The dependency on ArgoCD and Istio also creates more points of failure and maintenance.  

Do the benefits really outweigh the cost for every team and every use case? What good are great features on a product nobody wants to go through the struggle of deploying?

There is elegance in simplicity.

**Comment by berglh** [comment]
@thesuperzapper @AndersBennedsgaard @mnasty 

I'd echo the sentiments here about ArgoCD. I'm not here to argue the utility or widespread use of ArgoCD, most organisations I would expect have a centralised management of Kubernetes infrastructure. I currently work for an organisation where medium-sized teams are super-siloed, and therefore the scale of Kubernetes clusters is much smaller and managed by each team in their own cloud environments independently - none of them really yet have the scale to leverage the benefits of ArgoCD.

Our context is that we're operating a data lake on AWS. Our Kubernetes cluster has two primary functions;

1. We run Airflow to execute schedule DAGs (Primary Apache Spark), which are either managed by our team, or managed by customers, which are other groups in our organisation.  Airflow is still currently the best tool for us, even though Argo Workflows shows some promise, it has been somewhat lacking in official multi-tenancy support. Added to this, we have a very deep collection of datasets implemented in Airflow and porting them to a new workflow tool would be laboursome.

2. We use Kubeflow, primarily for the Notebook servers, where customers are split into namespaces/Kubeflow Workspaces and assume their roles via IRSA in EKS. From there, they start Apache Spark Kubernetes clusters,  slurp data from Glue/Athena tables and make data babies.

Our current EKS cluster and Airflow is deployed using Gitlab CI and Terraform. For better or worse, this suits us fine. Kubeflow on the other hand has historical been a mess of a kustomized helm chart and ansible - and I'm looking to ideally port this over to a Terraform based deployment.

Terraform is attractive in the sense it allows us to bundle all our AWS resources and our Kubernetes deployment into a single Terraform project. It's a Hashicorp shaped hammer I'm running around with.

Because of our rather simple configuration, introducing ArgoCD as another deployment we have to manage that is only performing a single function for us - deploying deployKF, is additional overhead that seems hard to justify despite the advantages presented.

In terms of the stateful application of manifests; Terraform is capable of managing the applying and deletion of manifests using the kubectl_manifests resource. If a manifest is deleted from the source repository, Terraform can delete that manifest from the cluster, which solves point 1.

The challenge is getting the correct order of these things applied, and as high-lighted; ensuring we are waiting on components being online before continuing - not so easy with Terraform, but potentially possible. I would be trading the ArgoCD automation for handling this logic myself which will create overhead in the long run having to adapt it to the changes of ArgoCD deployment.

I do see tremendous value in using deployKF to manage the generation of our deployment using the single YAML configuration as well as being cloud agnostic as Azure is on the horizon for us. If I could complete the manifest generation in a GitlabCI build phase, and then create the appropriate Terraform code to deploy the manifests - that for us would be ideal. If I were to understand the order that the app of apps operates in, then implementing the logic of the deployment wouldn't be all that difficult.

Where this strategy falls short is in the syncing aspect (as with the OIDC secrets mentioned), and this is something I'd need to understand better as it would be the only blocker in my mind.

Anyway, that's my two cents - I guess my lack of understanding on the "syncing" problem is likely where I'm coming undone. The other problem I have, is I'm on a bit of a ticking timebomb with our Kubeflow deployment which is ancient and failing on newer version of EKS. I'm not enamoured by how abandoned/sporadic the AWS Kubeflow project seems - so having to also learn ArgoCD is going to push out our horizon on migration of/to Kubeflow/deployKF, despite the fact it would save us time when considering other deployment methods.

**Comment by thesuperzapper** [comment]
@berglh thanks for the interesting write up.

It sounds like you will really benefit from deployKF, it will make your life a lot easier in maintaining Kubeflow. My recommendation (if you want to use deployKF in the short term), is to just treat the ArgoCD as "part of the platform" even if you never interact with it outside of the ["argocd sync script"](https://github.com/deployKF/deployKF/blob/main/scripts/sync_argocd_apps.sh) that we provide.

For example, you can trigger the sync script from a Terraform (or other CI/CD job), just pretend it's some external vendor's system you are triggering a "rollout" on.

___PS:__ there is lots of interesting stuff coming around Apache Airflow + deployKF, it's a little way off, but will be worth the wait!_

___PSS:__ if you currently deploy Airflow on Kubernetes, there is a good chance you are already using [our helm chart](https://github.com/airflow-helm/charts) to do it, and we think a lot of users will benefit from managing their "Airflow on Kubernetes" deployments under their deployKF configs and auth._

___PSSS:__ the "Kubeflow on AWS" distribution is [not maintained anymore](https://github.com/awslabs/kubeflow-manifests/issues/794)._

**Comment by Zeratoxx** [comment]
Hmm, no comment on K8s Operators? 🤔
https://github.com/deployKF/deployKF/issues/110#issuecomment-2000457452

**Comment by berglh** [comment]
> It sounds like you will really benefit from deployKF, it will make your life a lot easier in maintaining Kubeflow. My recommendation (if you want to use deployKF in the short term), is to just treat the ArgoCD as "part of the platform" even if you never interact with it outside of the ["argocd sync script"](https://github.com/deployKF/deployKF/blob/main/scripts/sync_argocd_apps.sh) that we provide.

The more I'm thinking about deployKF and the alternatives, this does make sense.
 
> _**PSS:** if you currently deploy Airflow on Kubernetes, there is a good chance you are already using [our helm chart](https://github.com/airflow-helm/charts) to do it, and we think a lot of users will benefit from managing their "Airflow on Kubernetes" deployments under their deployKF configs and auth._

I fell into the Astronomer fork via the Airflow docs. I've already banged my head on a few issues, such as not being able to sync secret updates via the Helm chart values (need to manually manage these); no Celery support for Elasticache Redis clusters is really *awesome* - not Airflow's fault, but there is some interesting options that Celery does support as a backend that are not currently available in Airflow [such as SQS](https://github.com/apache/airflow/issues/22863#issuecomment-1252313538). Rather annoying when I'm shooting for HA using affordable AWS services. 

Having deployKF manage this deployment as well would greatly simplify the stack, there is also an opportunity to wrap a single SAML service provider for every major component of our data lake when this comes to fruition, although, the inability of Kubeflow to handle workspace membership via SAML group attributes is a little sad. This issue has created significant administrative overhead for my team having to manage a separate user role mappings repository historically, which would be mostly distilled down to a group mappings repository with delegated AD group management to workspace owners/mangers if this capability existed.

> _**PSSS:** the "Kubeflow on AWS" distribution is [not maintained anymore](https://github.com/awslabs/kubeflow-manifests/issues/794)._

@thesuperzapper I will say, your experience and high level of contributions in the Airflow and Kubeflow space is certainly reassuring in terms of the strategic direction of this project. One thing that concerns me is that almost all, if not every commit in this repo is made by you. I'm not insinuating that you wouldn't accept PRs that provide merit to the project, but from a continuity perspective - that is one aspect of this project that gives me pause for concern, despite you offering paid support. 

I apologise this is starting to get a little off-topic, and I'm certainly not trying to make you *convince* me that [this is the way](https://i.giphy.com/6UFgdU9hirj1pAOJyN.webp). I understand the motivation behind this project given the lack of other suitable options.

I need to reconcile the risks of each option before I settle on a path forward as in my current multi-faceted role I have many mediocre-at-best hats I need to wear as Systems Programmer, Data/ML Engineer, Data Analyst, Data Domain Expert, Apache Spark SME/Trainer, On-prem Elastic Cloud Enterprise SME & Operator, Cloud Infrastructure Operations etc - I'm spread pretty thin and don't have the luxury of time to experiment and iterate frequently on the data lake stack.

**Comment by thesuperzapper** [comment]
> @thesuperzapper I will say, your experience and high level of contributions in the Airflow and Kubeflow space is certainly reassuring in terms of the strategic direction of this project. One thing that concerns me is that almost all, if not every commit in this repo is made by you. I'm not insinuating that you wouldn't accept PRs that provide merit to the project, but from a continuity perspective - that is one aspect of this project that gives me pause for concern, despite you offering paid support.

@berglh Your concerns are valid, and we are working to increase the number of contributors to deployKF so that it becomes a sustainable industry standard. By using and contributing to the project you will help this become a reality.

Also, in the near future we will allow sponsoring the project directly, and in exchange those sponsors will get special features, extra tools, and influence in what we develop. With these contributions, we plan to hire a number of full-time engineers that will ensure the continuity and velocity of the project.

To clarify, deployKF is and __always will be__ an open-source project, no payment/sponsorship will be needed to use deployKF to deploy and manage Kubeflow.

**Comment by suse-coder** [comment]
Please make a timoni chart (uses cuelang and has a helm adapter) (as flux-aio was also from the same developer developed). 

https://timoni.sh/quickstart/
Here would be a template: https://github.com/stefanprodan/podinfo/tree/master/timoni/podinfo
